ccm_web/client/src/pages/BulkSectionCreate.tsx

```diff
@@ -239,27 +329,30 @@ function BulkSectionCreate (props: BulkSectionCreateProps): JSX.Element {
       if (hasHeader(lines)) {
         lines = lines.slice(1)
       }
-
-      const rowInvalidations: SectionsRowInvalidation[] = []
-
-      const duplicateNamesInFileValidator: SectionRowsValidator = new DuplicateSectionInFileSectionRowsValidator()
-      rowInvalidations.push(...duplicateNamesInFileValidator.validate(lines))
-
-      const duplicatesNamesInCanvasValidator: DuplicateExistingSectionRowsValidator = new DuplicateExistingSectionRowsValidator()
-      rowInvalidations.push(...duplicatesNamesInCanvasValidator.validate(lines))
-
-      if (rowInvalidations.length !== 0) {
-        handleRowLevelInvalidationError(rowInvalidations)
-        return
-      }
-
       handleParseSuccess(lines)
     }).catch(e => {
       // TODO Not sure how to produce this error in real life
       handleSchemaError([{ error: 'Error processing file', type: InvalidationType.Error }])
     })
   }
 
+  const doClientValidation = (): SectionsRowInvalidation[] => {
+    const rowInvalidations: SectionsRowInvalidation[] = []
+
+    const duplicateNamesInFileValidator: SectionRowsValidator = new DuplicateSectionInFileSectionRowsValidator()
+    rowInvalidations.push(...duplicateNamesInFileValidator.validate(sectionNames))
+
+    return rowInvalidations
+  }
+
+  const doServerValidation = (): SectionsRowInvalidation[] => {
+    const rowInvalidations: SectionsRowInvalidation[] = []
+    const duplicatesNamesInCanvasValidator: DuplicateExistingSectionRowsValidator = new DuplicateExistingSectionRowsValidator()
+    rowInvalidations.push(...duplicatesNamesInCanvasValidator.validate(sectionNames))
+
+    return rowInvalidations
+  }
```

Comment:
One validates against sections that exist in the local file, one validates against duplicates that exist in  canvas.  They could be designed to use a single validator by making the validator itself more complicated but I chose to do it this way

Reply:
Ah, I'll be honest, I didn't look too closely at the exact implementations until this morning. Some similarities, but I take your point that it would make it more complicated. I'm okay keeping them separate.

That said, I see you are using `existingSectionNames` from the component scope directly in the validator. Could we make that an additional method parameter for `validate` (or a constructor parameter, instance variable) and then move this validator to be with the others in `BulkSectionCreateValidators.ts`?